// @elog/plugin-adapter/dist/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));

// src/markdown/matter-markdown.ts
var _graymatter = require('gray-matter'); var _graymatter2 = _interopRequireDefault(_graymatter);
var _shared = require('@elog/shared');
function matterMarkdownAdapter(post) {
  var _a;
  let body = post.body;
  try {
    const properties = post.properties;
    const props = __spreadProps(__spreadValues({}, properties), {
      title: (_a = properties == null ? void 0 : properties.title) == null ? void 0 : _a.replace(/"/g, "")
      // 临时去掉标题中的引号，至少保证文章页面是正常可访问的
    });
    body = _graymatter2.default.stringify(body, props, { lineWidth: -1 });
  } catch (e) {
    _shared.out.warning(`\u3010${post.properties.title}\u3011Front matter \u751F\u6210\u5931\u8D25\uFF0C\u8BF7\u68C0\u67E5\u6587\u6863\u5C5E\u6027`, e.message);
    _shared.out.debug(e);
  }
  return body;
}

// src/markdown/markdown.ts
function markdownAdapter(post) {
  let { body } = post;
  return body;
}

// src/wiki/render.ts
var _marked = require('marked');
var _querystring = require('querystring');
var langMap = {
  javascript: "js",
  typescript: "js",
  java: "java",
  shell: "bash",
  html: "html",
  xml: "xml",
  yaml: "yml"
};
var WikiRenderer = class extends _marked.Renderer {
  constructor() {
    super();
  }
  paragraph(text) {
    return text + "\n\n";
  }
  html(html) {
    return html;
  }
  heading(text, level) {
    return "h" + level + ". " + text + "\n\n";
  }
  strong(text) {
    return "*" + text + "*";
  }
  em(text) {
    return "_" + text + "_";
  }
  del(text) {
    return "-" + text + "-";
  }
  codespan(text) {
    return "{{" + text + "}}";
  }
  blockquote(quote) {
    return "{quote}" + quote + "{quote}";
  }
  br() {
    return "\n";
  }
  hr() {
    return "----";
  }
  link(href, _title, text) {
    const arr = [href];
    if (text) {
      arr.unshift(text);
    }
    return "[" + arr.join("|") + "]";
  }
  list(body, ordered) {
    const arr = body.trim().split("\n").filter(Boolean);
    const type = ordered ? "#" : "*";
    return arr.map((line) => {
      const isSub = line.startsWith(type);
      return "\n" + type + (isSub ? "" : " ") + line;
    }).join("") + "\n\n";
  }
  listitem(body) {
    return body + "\n";
  }
  image(href) {
    return "!" + href + "!";
  }
  table(header, body) {
    return header + body + "\n";
  }
  tablerow(content) {
    return content + "\n";
  }
  tablecell(content, flags) {
    const type = flags.header ? "||" : "|";
    return type + content;
  }
  code(code, lang) {
    if (lang) {
      lang = lang.toLowerCase();
      lang = langMap[lang] || "none";
    }
    let config = {
      language: lang,
      borderStyle: "solid",
      theme: "default",
      // dark is good
      linenumbers: true,
      collapse: false
    };
    const lineCount = code.split("\n").length;
    if (lineCount > 20) {
      config.collapse = true;
    }
    const param = _querystring.stringify.call(void 0, config, "|", "=");
    return "{code:" + param + "}\n" + code + "\n{code}\n\n";
  }
};
var md2Wiki = (markdown) => {
  const wikiRenderer = new WikiRenderer();
  return _marked.marked.parse(markdown, { renderer: wikiRenderer });
};

// src/wiki/wiki.ts
function wikiAdapter(post) {
  const { body } = post;
  return md2Wiki(body);
}

// src/html/html.ts
var _highlightjs = require('highlight.js'); var _highlightjs2 = _interopRequireDefault(_highlightjs);
var _markdownit = require('markdown-it'); var _markdownit2 = _interopRequireDefault(_markdownit);
function htmlAdapterWithHighlight(post) {
  let { body_html, body } = post;
  if (body_html) {
    return body_html;
  } else {
    return new (0, _markdownit2.default)({
      html: true,
      xhtmlOut: true,
      breaks: true,
      linkify: true,
      typographer: true,
      highlight: function(code, lang) {
        const language = _highlightjs2.default.getLanguage(lang) ? lang : "plaintext";
        return _highlightjs2.default.highlight(code, { language }).value;
      }
    }).render(body);
  }
}
function htmlAdapter(post) {
  let { body_html, body } = post;
  if (body_html) {
    return body_html;
  } else {
    return new (0, _markdownit2.default)({
      html: true,
      xhtmlOut: true,
      breaks: true,
      linkify: true,
      typographer: true
    }).render(body);
  }
}






exports.htmlAdapter = htmlAdapter; exports.htmlAdapterWithHighlight = htmlAdapterWithHighlight; exports.markdownAdapter = markdownAdapter; exports.matterMarkdownAdapter = matterMarkdownAdapter; exports.wikiAdapter = wikiAdapter;
